File: April 2020/24 - LRU Cache/Solution.py
from collections import OrderedDict

class LRUCache:

    def __init__(self, capacity: int):
        self.__size = capacity
        self.__cache = OrderedDict()
        
    def get(self, key: int) -> int:
        if key not in self.__cache:
            return -1
        else:
            self.__cache.move_to_end(key) #most recent
            return self.__cache[key] # return value
        
    def put(self, key: int, value: int) -> None:
        if key not in self.__cache:
            if len(self.__cache) >= self.__size: 
                self.__cache.popitem(last=False)
        else:
            self.__cache.move_to_end(key)        
        
        self.__cache[key] = value

File: April 2020/24 - LRU Cache/Solution.cpp
class LRUCache {
    int capacity;
    list<pair<int, int>> lrulist;
    unordered_map<int, list<pair<int, int>>::iterator> hash;
    
    void moveToFront(int key, int val) {
        // erase and insert element to the front of list
        lrulist.erase(hash[key]);// O(1) since using iterator
        lrulist.push_front(make_pair(key, val));
        hash[key] = lrulist.begin();
    }
    
public:
    LRUCache(int capacity) {
        this->capacity = capacity;
    }
    
    int get(int key) {
        if(hash.find(key) == hash.end()) return -1;
        // move the key to front(most recent)
        int val = (*hash[key]).second;
        moveToFront(key, val);
        return val;
    }
    
    void put(int key, int value) {
        if(hash.find(key) != hash.end())
            moveToFront(key, value);
        else {
            lrulist.push_front(make_pair(key, value));
            hash[key] = lrulist.begin(); 
            if(hash.size() > capacity) {
                hash.erase(lrulist.back().first);// least recent
                lrulist.pop_back();
            }
        }
    }
};